Added shared memory multi-server feature

jkuro-tii opened the pull request:
<!--
    Copyright 2023 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

## Description of changes

The memsocket application has been enhanced with the following new features and updates:

    Shared memory protection:
        Shared memory is allocated by a custom driver on the host. At runtime, the driver verifies 
        the requesting VM's identity and maps only the memory regions it is authorized to access.
        Memory access permissions are automatically generated by Nix scripts.

    Host Support:
        The application can now run on the host in addition to virtual machines (VMs).

    Multi-Server Capability:
        Multiple memsocket instances operating in server mode are now supported.
        Each server instance is configured with a list of specific client IDs it serves.
        Multiple servers can run simultaneously on a single VM or the host.

    Client ID Exclusivity:
        Each client ID must be managed by only one server across the entire system.

    Configuration Updates:
        Renamed some Nix configuration options for clarity.

    Option Changes:
        Added long option aliases: --server (-s), --listen (-l), --client (-c), --host (-h), --force (-f)
        The -c and -s options have been swapped:
            -c: Now used for running in server mode.
            -s: Now used for running in client mode.
        A new -l option has been introduced for server mode to specify the list of client IDs to be serviced.   
        A new -h option has been introduced for indicating run on the host

Introduced a Nix framework for the automatic generation of configurations for servers (currently supporting audio and video) and their clients.
Enabled the transfer of gui data via shared memory from application VMs to designated gui-vm server.

### Type of Change
- [X] New Feature
- [ ] Bug Fix
- [ ] Improvement / Refactor

## Related Issues / Tickets

<!--
Link to GitHub issues or JIRA tickets (if any) that this PR addresses or is related to
-->

## Checklist

<!--
Please check [X] for all items that apply. Leave [ ] if an item does not apply, but you have considered it.
Note that none of these are strict requirements — they are intended to inform reviewers.
Completing this checklist shows that you value and respect their time and effort.
-->

- [X] Clear summary in PR description
- [X] Detailed and meaningful commit message(s)
- [X] Commits are logically organized and squashed if appropriate
- [X] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [X] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [X] Author has run `make-checks` and it passes
- [ X All automatic GitHub Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [X] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing Instructions

### Applicable Targets
- [ ] Orin AGX `aarch64`
- [ ] Orin NX `aarch64`
- [X] Lenovo X1 `x86_64`
- [X] Dell Latitude `x86_64`

### Installation Method
- [ ] Requires full re-installation
- [X] Can be updated with `nixos-rebuild ... switch`
- [X] Other: requires rebooting after updating with `nixos-rebuild ... switch`

### Test Steps To Verify:
<!--
Provide clear, simple step-by-step instructions to verify the functionality.
Please do not assume that readers know everything you currently know.
-->
1. Run gui enabled applications, e.g.: chrome, outlook, slack, zathura, teams, etc. Check if they are displaying properly.
2. Repeat in a loop: close and start the above applications. Check if they are running and displaying properly.

josa41 commented:
Nothing too obvious to comment in the code.
I did test this also with pulseaudio and got everything to work just fine.
During my testing i found two issues.
1. If the application socket (pulseaudio in my tests) is not available when memsocket starts then data does not reach the socket. Restarting memsocket helps for this or adding a delay for memsocket to start after the application.
2. Memsocket/kernel crashes when a second memsocket server is started from command line on the same VM with another id

I passed more details to Jarek in slack

mbssrc commented:
Any idea how this can or should be tested?

josa41 commented:
Tested briefly and audio works through shared mem.
Had some trouble with permissions but that is not related to this change.

leivos-unikie commented:
Tested again on Lenovo-X1

The bug is fixed.

bat tests pass

Monitored now all VMs with the test script while starting all apps and closing them. Number of file descriptors change as noticed before. Nothing suspicious there.

johannarautanen commented:
**Checked with native Orin AGX and NX**

- build can be done
- Ghaf booted up
- applications launched
- automation test cases were ok

Notes:
- so far the build size flashed to the memory stick has been about 7-8GB, with this PR the size is 11-12 GB

johannarautanen commented:
**Checked with crosscompile Orin AGX and NX**'

Issues:
- cannot create the build fro AGX or NX

jkuro-tii commented:
It seems cross compiling for ARM is broken. For some reasons during building for the ARM platform, the `/arch/x86/` directory is accessed. 

```
 Running phase: buildPhase
build flags: SHELL=/nix/store/0irlcqx2n3qm6b1pc9rsd2i8qpvcccaj-bash-5.2p37/bin/bash KDIR=/nix/store/ffc1c388l2dv6asbpkh5v1l7nr27hchc-linux-aarch64-unknown-linux-gnu-6.6.62-dev/lib/modules/6.6.62/build MODULEDIR=\$\(out\)/lib/modules/6.6.62/kernel/drivers/char CFLAGS_kvm_ivshmem.o=\"-DCONFIG_KVM_IVSHMEM_SHM_SLOTS=10\"
cp ../drivers/char/ivshmem/kvm_ivshmem.c .
cp ../drivers/char/ivshmem/kvm_ivshmem.h .
make -C /nix/store/ffc1c388l2dv6asbpkh5v1l7nr27hchc-linux-aarch64-unknown-linux-gnu-6.6.62-dev/lib/modules/6.6.62/build M=/build/source/module modules
make[1]: Entering directory '/nix/store/ffc1c388l2dv6asbpkh5v1l7nr27hchc-linux-aarch64-unknown-linux-gnu-6.6.62-dev/lib/modules/6.6.62/build'
/nix/store/ffc1c388l2dv6asbpkh5v1l7nr27hchc-linux-aarch64-unknown-linux-gnu-6.6.62-dev/lib/modules/6.6.62/source/Makefile:768: /nix/store/ffc1c388l2dv6asbpkh5v1l7nr27hchc-linux-aarch64-unknown-linux-gnu-6.6.62-dev/lib/modules/6.6.62/source/arch/x86/Makefile: No such file or directory

```

leivos-unikie commented:
I was testing https://github.com/tiiuae/ci-test-automation/pull/207 with this ghaf image and noticed that audio test fails on Lenovo-X1. Tried youtube and it does not output any sound. Ghaf Audio Control does not show any stream while running youtube video.

leivos-unikie commented:
> I was testing [tiiuae/ci-test-automation#207](https://github.com/tiiuae/ci-test-automation/pull/207) with this ghaf image and noticed that audio test fails on Lenovo-X1. Tried youtube in chrome, it does not output any sound. Ghaf Audio Control does not show any stream while running youtube video.

The problem seems to be in test automation not being able to use pulseaudio. That failing automated audio test rendered ghaf audio once dysfunctional but after reboot I was not able to reproduce that state.

josa41 commented:
> > I was testing [tiiuae/ci-test-automation#207](https://github.com/tiiuae/ci-test-automation/pull/207) with this ghaf image and noticed that audio test fails on Lenovo-X1. Tried youtube in chrome, it does not output any sound. Ghaf Audio Control does not show any stream while running youtube video.
> 
> The problem seems to be in test automation not being able to use pulseaudio. That failing automated audio test rendered ghaf audio once dysfunctional but after reboot I was not able to reproduce that state.

What do you do in the automated tests? If you use
`export PULSE_SERVER=tcp:audio-vm`
it does not work when audio through shm is enabled as whole pulseaudio communication goes through shm.
You can use:
`export PULSE_SERVER=unix:/run/memsocket/audio-client.sock`
instead. The export should be automatically added to user shells but maybe this needs some tweaking for autotest environments.

leivos-unikie commented:
> > > I was testing [tiiuae/ci-test-automation#207](https://github.com/tiiuae/ci-test-automation/pull/207) with this ghaf image and noticed that audio test fails on Lenovo-X1. Tried youtube in chrome, it does not output any sound. Ghaf Audio Control does not show any stream while running youtube video.
> > 
> > 
> > The problem seems to be in test automation not being able to use pulseaudio. That failing automated audio test rendered ghaf audio once dysfunctional but after reboot I was not able to reproduce that state.
> 
> What do you do in the automated tests? If you use `export PULSE_SERVER=tcp:audio-vm` it does not work when audio through shm is enabled as whole pulseaudio communication goes through shm. You can use: `export PULSE_SERVER=unix:/run/memsocket/audio-client.sock` instead. The export should be automatically added to user shells but maybe this needs some tweaking for autotest environments.

Thanks Jon. I will try with that.

Anyway this problem with ci-test-automation is not a bug in ghaf. Just need to adapt test automation to the change. And the audio tests were not even yet integrated to the regular test runs. I remove the x1 bug tag.

leivos-unikie commented:
> > > > I was testing [tiiuae/ci-test-automation#207](https://github.com/tiiuae/ci-test-automation/pull/207) with this ghaf image and noticed that audio test fails on Lenovo-X1. Tried youtube in chrome, it does not output any sound. Ghaf Audio Control does not show any stream while running youtube video.
> > > 
> > > 
> > > The problem seems to be in test automation not being able to use pulseaudio. That failing automated audio test rendered ghaf audio once dysfunctional but after reboot I was not able to reproduce that state.
> > 
> > 
> > What do you do in the automated tests? If you use `export PULSE_SERVER=tcp:audio-vm` it does not work when audio through shm is enabled as whole pulseaudio communication goes through shm. You can use: `export PULSE_SERVER=unix:/run/memsocket/audio-client.sock` instead. The export should be automatically added to user shells but maybe this needs some tweaking for autotest environments.
> 
> Thanks Jon. I will try with that.
> 
> Anyway this problem with ci-test-automation is not a bug in ghaf. Just need to adapt test automation to the change. And the audio tests were not even yet integrated to the regular test runs. I remove the x1 bug tag.

Turned out that I was able to run automated audio tests when I added sudo to the commands but it works only once if the test setup does log in (before the tests) and log out (afterwards). Tested also manually that I can run the pulseaudio commands with sudo as long as I don't do Log out and Log in.  

If doing log in, log out and log in after boot, I get this (which would still work after first login):
[ghaf@chrome-vm:/tmp]$ sudo pactl list sources
[sudo] password for ghaf: 
Connection failure: Connection refused
pa_context_connect() failed: Connection refused

Clearly Log out / Log in changes something in the audio compared to the first login after boot.

jkuro-tii commented:
I noticed that behavior on my site. I'm working on that.

josa41 commented:
With audio its good to check audio control application if there are new streams when you open an app. This would tell if the app is able to connect to audiovm. At least slack is run on chrome browser. Also the configuration is exactly the same as with chrome vm so in a way it should be the same. The slight difference is that the vms are started in "random" order which could make slight difference although it should not matter (shm can handle that).

jkuro-tii commented:
I pushed a new version, which fixes problems when logging out and in.

jkuro-tii commented:
I pushed the version which uses shared memory only for sending video. Shared memory audio is temporarily disabled in order to figure out proper audio config data.

leivos-unikie commented:
I tested on x1 that all dekstop / app launcher gui buttons work.

Got once ghaf chrashing at login screen. After reboot I didn't have any audio devices working although previously they were working.

leivos-unikie commented:
But after another reboot the audio devices are back.

jkuro-tii commented:
It the last push I turned off sending audio through shared memory. So audio issues are unlikely related with this PR.

leivos-unikie commented:
> It the last push I turned off sending audio through shared memory. So audio issues are unlikely related with this PR.

Yes, I agree. 

Now when thinking what happened I believe that connecting to home wifi network with the same IP space than within ghaf virtual network might have caused the crash and also temporary loss of audio.

jkuro-tii commented:
Created [issue](https://github.com/tiiuae/ghaf/issues/973) regarding the problem with cross-building for the ARM64 platform.

jkuro-tii commented:
I added some fixes which allow using shared memory for gui only for Lenovo X11 platform. Problems with too big image, and to much memory consumption during building no longer exist.

Please review.

johannarautanen commented:
**Checked with crosscompile Orin AGX and NX**
Working:
- builds can be done
- Ghaf booted up and ghaf-screen visible on the external display
- applications launched
- automation test cases were ok
- build size was for NX about 7.9 and AGX 8.3GB when flashed to the USB memory stick

leivos-unikie commented:
That old bug "VM apps (running outside gui.vm) don't launch after log out and log in" is back again although it was fixed at some point. (See comments on Jan 13 - 14.)

johannarautanen commented:
**Checked with native Orin AGX and NX**
Working:

- builds can be done
- Ghaf booted up and ghaf-screen visible on the external display
- applications launched
- automation test cases were ok

vadika commented on modules/reference/personalize/authorizedSshKeys.nix:
```diff
@@ -28,7 +28,7 @@
     "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIMsRi+yikYJjpLhJZHlCb9N4Wb7fHUEh0Y6XekD9DyTQ juanpablo.ruiz@unikie.com"
     "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIEA7p7hHPvPT6uTU44Nb/p9/DT9mOi8mpqNllnpfawDE tanel@nixos"
-
+    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIIwGPH/oOrD1g15uiPV4gBKGk7f8ZBSyMEaptKOVs3NG jaroslawkurowski@TII-JaroslawKurowski"
```
Lets move it to separate commit?

jkuro-tii replied:
OK

jkuro-tii commented:
> That old bug "VM apps (running outside gui.vm) don't launch after log out and log in" is back again although it was fixed at some point. (See comments on Jan 13 - 14.)

The bug is related to the [PR](https://github.com/tiiuae/ghaf/issues/1036) I have raised. Even after fixing the issue by removing Waypipe tmp files, quickly logging out and back in still blocks services. I can provide a temporary workaround, but I believe fixing the root cause in the mentioned PR is the better approach

jkuro-tii commented:
Added removal waypipe of /tmp sockets in order to fix problem with re-logging (though the [problem with quick re-logging](https://github.com/tiiuae/ghaf/issues/1036) still persists).
Removed adding public ssh key.

jkuro-tii commented:
Fixed the logout process. A recent [update](https://github.com/tiiuae/ghaf/issues/1036) on the main branch caused a SIGKILL signal to be sent, preventing the proper propagation of the termination event

brianmcgillion commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -51,6 +108,99 @@ in
+              "gala-vm"
+              "zathura-vm"
+              "comms-vm"
+              "business-vm"
```
this should be set in the modules/reference/profiles/mvp. because it it not known here in common that these are the enabled vms.

brianmcgillion commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -51,6 +108,99 @@ in
+              "chrome-vm"
+              "business-vm"
+              "comms-vm"
+              "gala-vm"
```
same as above

brianmcgillion commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -139,9 +357,9 @@ in
+      (mkIf cfg.enable {
         environment.systemPackages = [
-          (pkgs.memsocket.override { vms = cfg.instancesCount; })
+          (pkgs.callPackage ../../../packages/pkgs-by-name/memsocket/package.nix { inherit (cfg) shmSlots; })
```
do not do this. use the  `(pkgs.memsocket.override { vms = cfg.shmSlotst; })` method to use the package

jkuro-tii replied:
Done.

jkuro-tii committed:
Shared memory refactoring. Added shm security mechanism

Signed-off-by: Jaroslaw Kurowski <jaroslaw.kurowski@tii.ae>

milva-unikie commented:
Please complete the 'Instructions for Testing' section, preferably using the new PR template.

```
## Testing Instructions

### Applicable Targets
- [ ] Orin AGX `aarch64`
- [ ] Orin NX `aarch64`
- [ ] Lenovo X1 `x86_64`
- [ ] Dell Latitude `x86_64`

### Installation Method
- [ ] Requires full re-installation
- [ ] Can be updated with `nixos-rebuild ... switch`
- [ ] Other: 

### Test Steps To Verify:
<!--
Provide clear, simple step-by-step instructions to verify the functionality.
Please do not assume that readers know everything you currently know.
-->
1. ...

```

milva-unikie commented:
**Tested on Lenovo-X1** (`nixos-rebuild switch`)

Rebased to the latest main 9c33f5e

- Apps are displayed properly
- Apps work even after repeatedly closing and opening them
- Audio & cameras work
- Test-automation passes

mbssrc reviewed (approved):
Suggest we merge this, and do some cleanups of the interface later
